aptos-move/mvhashmap/src/versioned_aggregators.rs

```diff
@@ -0,0 +1,708 @@
+                    })
+                    .flatten();
+
+                o.insert(CachePadded::new(Estimate(bypass)));
+            },
+            Entry::Vacant(_) => unreachable!("Versioned entry must exist when marking as estimate"),
+        };
+    }
+
+    fn remove(&mut self, txn_idx: TxnIndex) {
+        let deleted_entry = self.versioned_map.remove(&txn_idx);
+        // Entries should only be deleted if the transaction that produced them is
+        // aborted and re-executed, but abort must have marked the entry as an Estimate.
+        assert_matches!(
+            &*deleted_entry.expect("Entry must exist to be removed"),
+            AggregatorEntry::Estimate(_),
+            "Removed entry must be an Estimate",
+        );
+        // Incarnation changed output behavior, disable reading through estimates optimization.
+        self.read_estimate_deltas = false;
+    }
+
+    fn insert(&mut self, txn_idx: TxnIndex, entry: AggregatorEntry) {
+        use AggregatorEntry::*;
+
+        assert!(
+            !matches!(entry, Estimate(_)),
+            "Inserting Estimate is not allowed - must call mark_estimate"
+        );
+
+        match self.versioned_map.entry(txn_idx) {
+            Entry::Occupied(mut o) => {
+                if !match (&**o.get(), &entry) {
+                    // These are the cases where the transaction behavior with respect to the
+                    // aggregator may change (based on the information recorded in the Estimate).
+                    (Estimate(Some((delta_l, Some(id_l)))), Snapshot(id_r, delta_r)) => {
+                        *id_l == *id_r && *delta_l == *delta_r
+                    },
+                    (Estimate(Some((delta_l, None))), Delta(delta_r) | Value(_, Some(delta_r))) => {
+                        *delta_l == *delta_r
+                    },
+                    (Estimate(Some((_, None))), Snapshot(_, _)) => unreachable!(
+                        "Storing snapshot for aggregator ID \
+			 that previously contained a delta"
+                    ),
+                    // There was a value without fallback delta bypass before and still.
+                    (Estimate(None), Value(_, None)) => true,
+                    // Bypass stored in the estimate does not match the new entry.
+                    (Estimate(_), _) => false,
+
+                    // The following two cases are acceptable uses to record a value after txn
+                    // materialization / commit, as the value will never change.
+                    //
+                    // value & value pattern is allowed to not be too restrictive to the caller.
+                    //
+                    // The patterns ensure to avoid panic in the unreachable branch below, and
+                    // returning 'true' ensures that the bypass enabling logic is not affected.
+                    (Value(val_l, None), Value(val_r, _)) if val_l == val_r => true,
```

Comment:
Oh, right, before commit time the entry shouldn't be Occupied, and at the commit time if the last store value were different, it would be a critical correctness bug.